### PR TITLE
Feature: Enable logging to Weights & Biases

### PR DESCRIPTION
**Pull request by theyorubayesian**
- W&B is now initialized by default at the top of the training script (src/trainer.py). This can be made more flexible using the `report_to` CLI arg.  
- Modified metric writing functions to use either Tensorboard SummaryWriter or W&B logger.

### CODE REVIEW

File: src/trainer.py [hunk 1]
```diff
@@ -676,10 +717,11 @@ def group_texts(examples):
         num_proc=data_args.preprocessing_num_workers,
         load_from_cache_file=not data_args.overwrite_cache,
     )
-
+    
     # Enable tensorboard only on the master node
+    use_tensorboard = False
```
Comment (ToluClassics) [inline src/trainer.py:722]: Can we make this a parameter instead, as discusssed

File: train.sh [hunk 2]
```diff
@@ -6,6 +6,9 @@
 #SBATCH --account=def-kshook
 #SBATCH -o logs/train_afrit5_base_am_ha_sw.log
 
-nvidia-smi
+# nvidia-smi
+export WANDB_MODE="online"
+export WANDB_ENTITY="jarmy-naija"
+export WANDB_PROJECT="afriteva-v2"
```
Comment (ToluClassics) [inline train.sh:12]: Can we just add a comment to include this in case of public use

File: src/trainer.py [hunk 3]
```diff
@@ -725,14 +768,20 @@ def group_texts(examples):
...
-    num_train_steps = len(
-        tokenized_datasets["train"]) // train_batch_size * num_epochs
+    
+    # max_steps takes precedence over num_epochs
+    if training_args.max_steps > 0:
+        num_batches_per_epoch = (len(tokenized_datasets["train"]) // train_batch_size)
+        num_epochs = (training_args.max_steps // num_batches_per_epoch) + (training_args.max_steps % num_batches_per_epoch > 0)
+    else:
+        num_train_steps = len(
+        tokenized_datasets["train"]) // train_batch_size * training_args.num_epochs
```
Comment (ToluClassics) [inline src/trainer.py:784]: Can’t find where you initialise `  num_train_steps` in your if block

**Comment by theyorubayesian** [comment]
Updates:

- Initialized num_training_steps in the if block. 
- Allowed metric writing to work flexibly, controlled by training_arg: report_to 

Notes:
- summary_writer and wandb_run are initialized to None if report_to is not "tensorboard" or "wandb" respectively.
